### PR TITLE
Do not warn when invoking APIs that has SupportedOSPlatformGuard attribute

### PR DESCRIPTION
**Pull request by buyaa-n**
### Do not warn when invoking Guard APIs that has SupportedOSPlatformGuard attribute

Fixes issue @ManickaP encountered on https://github.com/dotnet/runtime/compare/main...ManickaP:runtime:quic-platform-guard?expand=1

 Added unit test that reproes the issue

Repro:
```cs
[assembly: System.Runtime.Versioning.SupportedOSPlatform(""windows"")]
[assembly: System.Runtime.Versioning.SupportedOSPlatform(""linux"")]
[assembly: System.Runtime.Versioning.SupportedOSPlatform(""osx"")]

namespace System.Net.Quic;
public sealed partial class QuicConnection
{
    [SupportedOSPlatformGuard("windows")]
    [SupportedOSPlatformGuard("linux")]
    [SupportedOSPlatformGuard("osx")]
    public static bool IsSupported => false;
    ...
}
```
The Guard API used from another assembly, should not warn on `QuicConnection.IsSupported` invocation
```cs
namespace System.Net.Http;
internal sealed partial class HttpConnectionPool : IDisposable
{
    private async ValueTask<QuicConnection> ConnectAsync(CancellationToken cancellationToken) ()
    {
        if (QuicConnection.IsSupported) // Currently warning here, fixing this
        {
            await QuicConnection.ConnectAsync(new QuicClientConnectionOptions(), cancellationToken).ConfigureAwait(false);
        }
    }
}
```

### CODE REVIEW

**Comment by buyaa-n** [comment]
@sharwell there is something broke in the repo, normal build succeeds but `dotnet build /t:pack` fails with MSBuild error: 
How can I fix this? cc @mavasani 
```log
Microsoft.CodeAnalysis.PublicApiAnalyzers.Package failed with 1 error(s) (1.0s)
    D:\dotnet\roslyn-analyzers\eng\GenerateAnalyzerNuspec.targets(113,5): error MSB3073: The command ""D:\dotnet\roslyn-analyzers\.dotnet\dotnet.exe" --roll-forward major "D:\dotnet\roslyn-analyzers\artifacts\bin\GenerateDocumentationAndConfigFiles\Debug\net9.0\GenerateDocumentationAndConfigFiles.dll" "-validateOnly:false" "D:\dotnet\roslyn-analyzers\artifacts\obj\Microsoft.CodeAnalysis.PublicApiAnalyzers.Package\Debug\netstandard2.0\Rulesets" "D:\dotnet\roslyn-analyzers\artifacts\obj\Microsoft.CodeAnalysis.PublicApiAnalyzers.Package\Debug\netstandard2.0\Editorconfig" "D:\dotnet\roslyn-analyzers\artifacts\obj\Microsoft.CodeAnalysis.PublicApiAnalyzers.Package\Debug\netstandard2.0\GlobalAnalyzerConfigs" "D:\dotnet\roslyn-analyzers\artifacts\bin\\" "Debug" "netstandard2.0" "Microsoft.CodeAnalysis.PublicApiAnalyzers.dll;Microsoft.CodeAnalysis.PublicApiAnalyzers.CodeFixes.dll" "D:\dotnet\roslyn-analyzers\artifacts\obj\Microsoft.CodeAnalysis.PublicApiAnalyzers.Package\Debug\netstandard2.0\Build" "Microsoft.CodeAnalysis.PublicApiAnalyzers.props" "D:\dotnet\roslyn-analyzers\artifacts\obj\Microsoft.CodeAnalysis.PublicApiAnalyzers.Package\Debug\netstandard2.0\Build" "Microsoft.CodeAnalysis.PublicApiAnalyzers.targets" "" "D:\dotnet\roslyn-analyzers\src\PublicApiAnalyzers" "Microsoft.CodeAnalysis.PublicApiAnalyzers.md" "D:\dotnet\roslyn-analyzers\src\PublicApiAnalyzers" "Microsoft.CodeAnalysis.PublicApiAnalyzers.sarif" "3.11.0" Microsoft.CodeAnalysis.PublicApiAnalyzers false true "" false" exited with code -2147450730.
  Roslyn.Diagnostics.Analyzers.Package failed with 1 error(s) (3.2s)
    D:\dotnet\roslyn-analyzers\eng\GenerateAnalyzerNuspec.targets(113,5): error MSB3073: The command ""D:\dotnet\roslyn-analyzers\.dotnet\dotnet.exe" --roll-forward major "D:\dotnet\roslyn-analyzers\artifacts\bin\GenerateDocumentationAndConfigFiles\Debug\net9.0\GenerateDocumentationAndConfigFiles.dll" "-validateOnly:false" "D:\dotnet\roslyn-analyzers\artifacts\obj\Roslyn.Diagnostics.Analyzers.Package\Debug\netstandard2.0\Rulesets" "D:\dotnet\roslyn-analyzers\artifacts\obj\Roslyn.Diagnostics.Analyzers.Package\Debug\netstandard2.0\Editorconfig" "D:\dotnet\roslyn-analyzers\artifacts\obj\Roslyn.Diagnostics.Analyzers.Package\Debug\netstandard2.0\GlobalAnalyzerConfigs" "D:\dotnet\roslyn-analyzers\artifacts\bin\\" "Debug" "netstandard2.0" "Roslyn.Diagnostics.Analyzers.dll;Roslyn.Diagnostics.CSharp.Analyzers.dll;Roslyn.Diagnostics.VisualBasic.Analyzers.dll" "D:\dotnet\roslyn-analyzers\artifacts\obj\Roslyn.Diagnostics.Analyzers.Package\Debug\netstandard2.0\Build" "Roslyn.Diagnostics.Analyzers.props" "D:\dotnet\roslyn-analyzers\artifacts\obj\Roslyn.Diagnostics.Analyzers.Package\Debug\netstandard2.0\Build" "Roslyn.Diagnostics.Analyzers.targets" "" "D:\dotnet\roslyn-analyzers\src\Roslyn.Diagnostics.Analyzers" "Roslyn.Diagnostics.Analyzers.md" "D:\dotnet\roslyn-analyzers\src\Roslyn.Diagnostics.Analyzers" "Roslyn.Diagnostics.Analyzers.sarif" "3.11.0" Roslyn.Diagnostics.Analyzers false true "" false" exited with code -2147450730.
  Microsoft.CodeAnalysis.ResxSourceGenerator.Package failed with 1 error(s) (1.3s)
    D:\dotnet\roslyn-analyzers\eng\GenerateAnalyzerNuspec.targets(113,5): error MSB3073: The command ""D:\dotnet\roslyn-analyzers\.dotnet\dotnet.exe" --roll-forward major "D:\dotnet\roslyn-analyzers\artifacts\bin\GenerateDocumentationAndConfigFiles\Debug\net9.0\GenerateDocumentationAndConfigFiles.dll" "-validateOnly:false" "D:\dotnet\roslyn-analyzers\artifacts\obj\Microsoft.CodeAnalysis.ResxSourceGenerator.Package\Debug\netstandard2.0\Rulesets" "D:\dotnet\roslyn-analyzers\artifacts\obj\Microsoft.CodeAnalysis.ResxSourceGenerator.Package\Debug\netstandard2.0\Editorconfig" "D:\dotnet\roslyn-analyzers\artifacts\obj\Microsoft.CodeAnalysis.ResxSourceGenerator.Package\Debug\netstandard2.0\GlobalAnalyzerConfigs" "D:\dotnet\roslyn-analyzers\artifacts\bin\\" "Debug" "netstandard2.0" "Microsoft.CodeAnalysis.ResxSourceGenerator.dll;Microsoft.CodeAnalysis.ResxSourceGenerator.CSharp.dll;Microsoft.CodeAnalysis.ResxSourceGenerator.VisualBasic.dll" "D:\dotnet\roslyn-analyzers\artifacts\obj\Microsoft.CodeAnalysis.ResxSourceGenerator.Package\Debug\netstandard2.0\Build" "Microsoft.CodeAnalysis.ResxSourceGenerator.props" "D:\dotnet\roslyn-analyzers\artifacts\obj\Microsoft.CodeAnalysis.ResxSourceGenerator.Package\Debug\netstandard2.0\Build" "Microsoft.CodeAnalysis.ResxSourceGenerator.targets" "" "D:\dotnet\roslyn-analyzers\src\Microsoft.CodeAnalysis.ResxSourceGenerator" "Microsoft.CodeAnalysis.ResxSourceGenerator.md" "D:\dotnet\roslyn-analyzers\src\Microsoft.CodeAnalysis.ResxSourceGenerator" "Microsoft.CodeAnalysis.ResxSourceGenerator.sarif" "3.11.0" Microsoft.CodeAnalysis.ResxSourceGenerator false true "true" false" exited with code -2147450730.
  Microsoft.CodeAnalysis.NetAnalyzers.Package failed with 1 error(s) (1.5s)
    D:\dotnet\roslyn-analyzers\eng\GenerateAnalyzerNuspec.targets(113,5): error MSB3073: The command ""D:\dotnet\roslyn-analyzers\.dotnet\dotnet.exe" --roll-forward major "D:\dotnet\roslyn-analyzers\artifacts\bin\GenerateDocumentationAndConfigFiles\Debug\net9.0\GenerateDocumentationAndConfigFiles.dll" "-validateOnly:false" "D:\dotnet\roslyn-analyzers\artifacts\obj\Microsoft.CodeAnalysis.NetAnalyzers.Package\Debug\netstandard2.0\Rulesets" "D:\dotnet\roslyn-analyzers\artifacts\obj\Microsoft.CodeAnalysis.NetAnalyzers.Package\Debug\netstandard2.0\Editorconfig" "D:\dotnet\roslyn-analyzers\artifacts\obj\Microsoft.CodeAnalysis.NetAnalyzers.Package\Debug\netstandard2.0\GlobalAnalyzerConfigs" "D:\dotnet\roslyn-analyzers\artifacts\bin\\" "Debug" "netstandard2.0" "Microsoft.CodeAnalysis.NetAnalyzers.dll;Microsoft.CodeAnalysis.CSharp.NetAnalyzers.dll;Microsoft.CodeAnalysis.VisualBasic.NetAnalyzers.dll" "D:\dotnet\roslyn-analyzers\artifacts\obj\Microsoft.CodeAnalysis.NetAnalyzers.Package\Debug\netstandard2.0\Build" "Microsoft.CodeAnalysis.NetAnalyzers.props" "D:\dotnet\roslyn-analyzers\artifacts\obj\Microsoft.CodeAnalysis.NetAnalyzers.Package\Debug\netstandard2.0\Build" "Microsoft.CodeAnalysis.NetAnalyzers.targets" "DisableNETAnalyzersForNuGetPackage.props" "D:\dotnet\roslyn-analyzers\src\NetAnalyzers" "Microsoft.CodeAnalysis.NetAnalyzers.md" "D:\dotnet\roslyn-analyzers\src\NetAnalyzers" "Microsoft.CodeAnalysis.NetAnalyzers.sarif" "9.0.0" Microsoft.CodeAnalysis.NetAnalyzers true true "" false" exited with code -2147450730.
  Microsoft.CodeAnalysis.Analyzers.Package failed with 1 error(s) (0.8s)
    D:\dotnet\roslyn-analyzers\eng\GenerateAnalyzerNuspec.targets(113,5): error MSB3073: The command ""D:\dotnet\roslyn-analyzers\.dotnet\dotnet.exe" --roll-forward major "D:\dotnet\roslyn-analyzers\artifacts\bin\GenerateDocumentationAndConfigFiles\Debug\net9.0\GenerateDocumentationAndConfigFiles.dll" "-validateOnly:false" "D:\dotnet\roslyn-analyzers\artifacts\obj\Microsoft.CodeAnalysis.Analyzers.Package\Debug\netstandard2.0\Rulesets" "D:\dotnet\roslyn-analyzers\artifacts\obj\Microsoft.CodeAnalysis.Analyzers.Package\Debug\netstandard2.0\Editorconfig" "D:\dotnet\roslyn-analyzers\artifacts\obj\Microsoft.CodeAnalysis.Analyzers.Package\Debug\netstandard2.0\GlobalAnalyzerConfigs" "D:\dotnet\roslyn-analyzers\artifacts\bin\\" "Debug" "netstandard2.0" "Microsoft.CodeAnalysis.Analyzers.dll;Microsoft.CodeAnalysis.CSharp.Analyzers.dll;Microsoft.CodeAnalysis.VisualBasic.Analyzers.dll" "D:\dotnet\roslyn-analyzers\artifacts\obj\Microsoft.CodeAnalysis.Analyzers.Package\Debug\netstandard2.0\Build" "Microsoft.CodeAnalysis.Analyzers.props" "D:\dotnet\roslyn-analyzers\artifacts\obj\Microsoft.CodeAnalysis.Analyzers.Package\Debug\netstandard2.0\Build" "Microsoft.CodeAnalysis.Analyzers.targets" "" "D:\dotnet\roslyn-analyzers\src\Microsoft.CodeAnalysis.Analyzers" "Microsoft.CodeAnalysis.Analyzers.md" "D:\dotnet\roslyn-analyzers\src\Microsoft.CodeAnalysis.Analyzers" "Microsoft.CodeAnalysis.Analyzers.sarif" "3.11.0" Microsoft.CodeAnalysis.Analyzers false true "" false" exited with code -2147450730.
  Microsoft.CodeAnalysis.PerformanceSensitiveAnalyzers.Package failed with 1 error(s) (0.6s)
    D:\dotnet\roslyn-analyzers\eng\GenerateAnalyzerNuspec.targets(113,5): error MSB3073: The command ""D:\dotnet\roslyn-analyzers\.dotnet\dotnet.exe" --roll-forward major "D:\dotnet\roslyn-analyzers\artifacts\bin\GenerateDocumentationAndConfigFiles\Debug\net9.0\GenerateDocumentationAndConfigFiles.dll" "-validateOnly:false" "D:\dotnet\roslyn-analyzers\artifacts\obj\Microsoft.CodeAnalysis.PerformanceSensitiveAnalyzers.Package\Debug\netstandard2.0\Rulesets" "D:\dotnet\roslyn-analyzers\artifacts\obj\Microsoft.CodeAnalysis.PerformanceSensitiveAnalyzers.Package\Debug\netstandard2.0\Editorconfig" "D:\dotnet\roslyn-analyzers\artifacts\obj\Microsoft.CodeAnalysis.PerformanceSensitiveAnalyzers.Package\Debug\netstandard2.0\GlobalAnalyzerConfigs" "D:\dotnet\roslyn-analyzers\artifacts\bin\\" "Debug" "netstandard2.0" "Microsoft.CodeAnalysis.PerformanceSensitiveAnalyzers.dll;Microsoft.CodeAnalysis.CSharp.PerformanceSensitiveAnalyzers.dll;Microsoft.CodeAnalysis.CSharp.PerformanceSensitiveAnalyzers.CodeFixes.dll" "D:\dotnet\roslyn-analyzers\artifacts\obj\Microsoft.CodeAnalysis.PerformanceSensitiveAnalyzers.Package\Debug\netstandard2.0\Build" "Microsoft.CodeAnalysis.PerformanceSensitiveAnalyzers.props" "D:\dotnet\roslyn-analyzers\artifacts\obj\Microsoft.CodeAnalysis.PerformanceSensitiveAnalyzers.Package\Debug\netstandard2.0\Build" "Microsoft.CodeAnalysis.PerformanceSensitiveAnalyzers.targets" "" "D:\dotnet\roslyn-analyzers\src\PerformanceSensitiveAnalyzers" "Microsoft.CodeAnalysis.PerformanceSensitiveAnalyzers.md" "D:\dotnet\roslyn-analyzers\src\PerformanceSensitiveAnalyzers" "Microsoft.CodeAnalysis.PerformanceSensitiveAnalyzers.sarif" "3.11.0" Microsoft.CodeAnalysis.PerformanceSensitiveAnalyzers false true "" false" exited with code -2147450730.
  Microsoft.CodeAnalysis.BannedApiAnalyzers.Package failed with 1 error(s) (0.2s)
    D:\dotnet\roslyn-analyzers\eng\GenerateAnalyzerNuspec.targets(113,5): error MSB3073: The command ""D:\dotnet\roslyn-analyzers\.dotnet\dotnet.exe" --roll-forward major "D:\dotnet\roslyn-analyzers\artifacts\bin\GenerateDocumentationAndConfigFiles\Debug\net9.0\GenerateDocumentationAndConfigFiles.dll" "-validateOnly:false" "D:\dotnet\roslyn-analyzers\artifacts\obj\Microsoft.CodeAnalysis.BannedApiAnalyzers.Package\Debug\netstandard2.0\Rulesets" "D:\dotnet\roslyn-analyzers\artifacts\obj\Microsoft.CodeAnalysis.BannedApiAnalyzers.Package\Debug\netstandard2.0\Editorconfig" "D:\dotnet\roslyn-analyzers\artifacts\obj\Microsoft.CodeAnalysis.BannedApiAnalyzers.Package\Debug\netstandard2.0\GlobalAnalyzerConfigs" "D:\dotnet\roslyn-analyzers\artifacts\bin\\" "Debug" "netstandard2.0" "Microsoft.CodeAnalysis.BannedApiAnalyzers.dll;Microsoft.CodeAnalysis.CSharp.BannedApiAnalyzers.dll;Microsoft.CodeAnalysis.VisualBasic.BannedApiAnalyzers.dll" "D:\dotnet\roslyn-analyzers\artifacts\obj\Microsoft.CodeAnalysis.BannedApiAnalyzers.Package\Debug\netstandard2.0\Build" "Microsoft.CodeAnalysis.BannedApiAnalyzers.props" "D:\dotnet\roslyn-analyzers\artifacts\obj\Microsoft.CodeAnalysis.BannedApiAnalyzers.Package\Debug\netstandard2.0\Build" "Microsoft.CodeAnalysis.BannedApiAnalyzers.targets" "" "D:\dotnet\roslyn-analyzers\src\Microsoft.CodeAnalysis.BannedApiAnalyzers" "Microsoft.CodeAnalysis.BannedApiAnalyzers.md" "D:\dotnet\roslyn-analyzers\src\Microsoft.CodeAnalysis.BannedApiAnalyzers" "Microsoft.CodeAnalysis.BannedApiAnalyzers.sarif" "3.11.0" Microsoft.CodeAnalysis.BannedApiAnalyzers false true "" false" exited with code -2147450730.
  Text.Analyzers.Package failed with 1 error(s) (0.6s)
    D:\dotnet\roslyn-analyzers\eng\GenerateAnalyzerNuspec.targets(113,5): error MSB3073: The command ""D:\dotnet\roslyn-analyzers\.dotnet\dotnet.exe" --roll-forward major "D:\dotnet\roslyn-analyzers\artifacts\bin\GenerateDocumentationAndConfigFiles\Debug\net9.0\GenerateDocumentationAndConfigFiles.dll" "-validateOnly:false" "D:\dotnet\roslyn-analyzers\artifacts\obj\Text.Analyzers.Package\Debug\netstandard2.0\Rulesets" "D:\dotnet\roslyn-analyzers\artifacts\obj\Text.Analyzers.Package\Debug\netstandard2.0\Editorconfig" "D:\dotnet\roslyn-analyzers\artifacts\obj\Text.Analyzers.Package\Debug\netstandard2.0\GlobalAnalyzerConfigs" "D:\dotnet\roslyn-analyzers\artifacts\bin\\" "Debug" "netstandard2.0" "Text.Analyzers.dll;Text.CSharp.Analyzers.dll;Text.VisualBasic.Analyzers.dll;C:\Users\bunamnan\.nuget\packages\humanizer.core\2.14.1\lib\netstandard1.0\Humanizer.dll" "D:\dotnet\roslyn-analyzers\artifacts\obj\Text.Analyzers.Package\Debug\netstandard2.0\Build" "Text.Analyzers.props" "D:\dotnet\roslyn-analyzers\artifacts\obj\Text.Analyzers.Package\Debug\netstandard2.0\Build" "Text.Analyzers.targets" "DisableNETAnalyzersForNuGetPackage.props" "D:\dotnet\roslyn-analyzers\src\Text.Analyzers" "Text.Analyzers.md" "D:\dotnet\roslyn-analyzers\src\Text.Analyzers" "Text.Analyzers.sarif" "3.11.0" Text.Analyzers true true "" false" exited with code -2147450730.

Build failed with 8 error(s) in 77.0s
```

**Review by ManickaP — APPROVED** [review]
Not an expert here though.

**Comment by buyaa-n** [comment]
> @sharwell there is something broke in the repo, normal build succeeds but dotnet build /t:pack fails with MSBuild error:
> How can I fix this? cc @mavasani

Turns out .NET version I was using was higher than the repo version, after the versions sync `dotnet build /t:pack` succeeded and produced the missing update

File: src/NetAnalyzers/RulesMissingDocumentation.md [hunk 1]
```diff
@@ -2,7 +2,5 @@
 
 Rule ID | Missing Help Link | Title |
 --------|-------------------|-------|
-CA1871 | <https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1871> | Do not pass a nullable struct to 'ArgumentNullException.ThrowIfNull' |
 CA2022 | <https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2022> | Avoid inexact read with 'Stream.Read' |
-CA2264 | <https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2264> | Do not pass a non-nullable value to 'ArgumentNullException.ThrowIfNull' |
```
Comment (carlossanlop) [inline src/NetAnalyzers/RulesMissingDocumentation.md:7]: Why were these removed?

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.GuardedCallsTests.cs [hunk 2]
```diff
@@ -1216,6 +1217,76 @@ static void M1()
...
+        {
+             Library.AMethod();
+        }
+    }
+}";
+            var test = SetupDependencyAndTestCSWithOneSourceFile(csCurrentAssemblyCode, csDependencyCode);
+            test.TestState.AnalyzerConfigFiles.Add(("/.editorconfig", $@"root = true
+
+[*]
+build_property.TargetFramework = net5
```
Comment (carlossanlop) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.GuardedCallsTests.cs:1254]: Would it make sense to make this test method a Theory and pass all the .NET versions that support this? Or is it not worth testing the code in all the .NET versions equal or greater than .NET 5?

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.GuardedCallsTests.cs [hunk 3]
```diff
@@ -1216,6 +1217,76 @@ static void M1()
...
+public class Library
+{
+    [System.Runtime.Versioning.SupportedOSPlatformGuard(""windows"")]
+    [System.Runtime.Versioning.SupportedOSPlatformGuard(""linux"")]
+    public static bool IsSupported => false;
+
+    public static void AMethod() { }
+}";
+            csDependencyCode = @$"[assembly: System.Runtime.Versioning.SupportedOSPlatform(""windows"")]
+                                  [assembly: System.Runtime.Versioning.SupportedOSPlatform(""linux"")]
```
Comment (carlossanlop) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.GuardedCallsTests.cs:1233]: I see that the helper method below uses `"/PreviewAssembly/AssemblyInfo.g.cs"` to place the SupportedOSPlatform attributes for the assembly.

Would it make sense to have an extra test where the SupportedOSPlatform and the SupportedOSPlatformGuard attributes are within the same file?

What about a test case where the SupportedOSPlatform attribute is in a regular file (not a *.g.cs file)?

**Review by carlossanlop — APPROVED** [review]
The source changes look good, but I wonder if we should have additional test cases.

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.GuardedCallsTests.cs [hunk 4]
```diff
@@ -1216,6 +1217,76 @@ static void M1()
...
+                                  {csDependencyCode}";
+
+            string csCurrentAssemblyCode = @"
+using System;
+
+public class Program
+{
+    public void ProgramMethod()
+    {
+        [|Library.AMethod()|]; // Not guarded, warns
```
Comment (carlossanlop) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.GuardedCallsTests.cs:1243]: Can you add a test case where it _is_ guarded?

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs [hunk 5]
```diff
@@ -1407,14 +1407,18 @@ void CheckOperationAttributes(ISymbol symbol, bool checkParents)
                             platformSpecificOperations.TryAdd(new KeyValuePair<IOperation, ISymbol>(operation, symbol), (notSuppressedAttributes, callSiteAttributes.Platforms));
                         }
                     }
-                    else if (TryCopyAttributesNotSuppressedByMsBuild(operationAttributes.Platforms!, msBuildPlatforms, out var copiedAttributes))
+                    else if (!OperationHasOnlyAssemblyAttributesAndCalledFromSameAssembly(operationAttributes, symbol, containingSymbol) &&
+                             TryCopyAttributesNotSuppressedByMsBuild(operationAttributes.Platforms!, msBuildPlatforms, out var copiedAttributes))
```
Comment (carlossanlop) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs:1411]: I see this is for assembly only. What about the case where you have a partial type split in many files, and SupportedOSPlatform decorates the type, and that type has a nested type? Is that a case that would trigger a similar error?

FileOne.cs
```cs
[SupportedOSPlatform("...")]
public partial class MyPartialClass
{
    /// some code
}
```

FileTwo.cs
```cs
public partial class MyPartialClass
{
    /// some code
    public class MyNestedClass
    {
        [SupportedOsPlatformGuard("...")]
        public bool IsSupported => true;
    }
}
```

I don't know if this case would make sense, but I was just wondering if there were more cases to consider.

File: src/NetAnalyzers/RulesMissingDocumentation.md (hunk 1 above)
Comment (buyaa-n) [reply src/NetAnalyzers/RulesMissingDocumentation.md:7]: Because those docs are now added, and the CI would complain to run `msbuild /t:pack` to remove these from `RulesMissingDocumentation.md`

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.GuardedCallsTests.cs (hunk 2 above)
Comment (buyaa-n) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.GuardedCallsTests.cs:1254]: So, the analyzer bails out when project TargetFramework is below .NET 5.0 or missing, that is why it is added here, there are existing tests that test this scenario

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.GuardedCallsTests.cs (hunk 4 above)
Comment (buyaa-n) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.GuardedCallsTests.cs:1243]: That is just below:
```cs
        if (Library.IsSupported) // this is the guard
        {
             Library.AMethod();
        }
```

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.GuardedCallsTests.cs (hunk 3 above)
Comment (buyaa-n) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.GuardedCallsTests.cs:1233]: > I see that the helper method below uses "/PreviewAssembly/AssemblyInfo.g.cs" to place the SupportedOSPlatform attributes for the assembly.

> Would it make sense to have an extra test where the SupportedOSPlatform and the SupportedOSPlatformGuard attributes are within the same file?

Well, there is plenty of tests, just all the other tests for this scenario are added within a same assembly, that is why this bug did not caught with them 

> What about a test case where the SupportedOSPlatform attribute is in a regular file (not a *.g.cs file)?

The attributes can be in any `*.cs` file of the assembly, it is just kind of convention that assembly level attributes added into `AssemblyInfo.cs`

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs (hunk 5 above)
Comment (buyaa-n) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs:1411]: I believe those split source will be parsed to one NamedSymbol, therefore there will be no any difference

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.GuardedCallsTests.cs (hunk 3 above)
Comment (buyaa-n) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzerTests.GuardedCallsTests.cs:1239]: Thank you for the review, as suggested I added more scenarios like `UnsupportedOSPlatformGuard` attribute used for guard, the Guard property accessing internal field (which was causing a warning previously which fix resulted this bug).

**Comment by buyaa-n** [comment]
> The source changes look good, but I wonder if we should have additional test cases.

Added additional tests, thanks!

**Comment by ManickaP** [comment]
@buyaa-n do you know when this will be available in runtime repo? Or could you let me know when the time comes?

**Comment by buyaa-n** [comment]
Might be it https://github.com/dotnet/runtime/pull/103473

**Comment by ManickaP** [comment]
Work like charm, thank you!
